### PR TITLE
✨ Is705/api keys with expiration

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

- Can create API-keys with expiration (optional) via the web-server's API ``0.8.0 → 0.9.0``
image
- Adds periodic background task to prune api-keys that reached their expiration date

### devops
 - New env-var ``GARBAGE_COLLECTOR_PRUNE_APIKEYS_INTERVAL_S`` to configure pruning period. Defaults to 1h.
 
 
## Related issue/s

- closes ITISFoundation/osparc-issues#705

## How to test

```cmd
cd services/web/server
make install-dev
pytest -vv tests/unit/**/test_login_api_keys.py
```

## Checklist

- [x] add to to GC
- [x] extend models in OAS
- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [ ] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/garbage_collector_tasks_api_keys.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        if deleted := await prune_expired_api_keys(app):
+
+            # broadcast force logout of user_id
+            for api_key in deleted:
+                logger.info("API-key %s expired and was removed", f"{api_key=}")
+
+        else:
+            logger.info("No API keys expired")
+
+        await asyncio.sleep(wait_period_s)
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/garbage_collector_tasks_api_keys.py:46]: We already have a discussion about this pattern [here](https://github.com/ITISFoundation/osparc-simcore/pull/3380#discussion_r979593787) so let's finish it up there.

**Review by GitHK — COMMENTED** [review]
👍 Looking good.
Please have a look at the discussion linked below.

**Review by mrnicegyu11 — CHANGES_REQUESTED** [review]
Please add:
- DevOps Changes Warning clearly in the PR title
- Instruction to add a graylog alert for `Failed to delete API key %s. Ignoring error` in the devops section of this PR
- Instruction that the signup-link generation admin-panel needs to be changed to use webserver-API routes in the devops section

If possible: Please also add a env-var config option to the GC, by default false, to disable expired API key deletion entirely from the GC, if possible.

Thanks.

Question: Will the frontend be able to display a clear "Your account is expired (please purchase more oSparc tokens)" error using this backend logic? I guess that this is at least the intuitive user-story.

**Comment by pcrespov** [comment]
> * DevOps Changes Warning clearly in the PR title

I did not add it in the title because it does not require any change in the current setup. 
I added a section for devops to inform about the definition of a new variable. I guess the icon I used (⚠️ devops) is a bit misleading. I should perhaps just write "devops"

> * Instruction to add a graylog alert for `Failed to delete API key %s. Ignoring error` in the devops section of this PR
IMO there is no alert necessary here (for the moment). In any case, let me know your thoughts about these alerts.

> * Instruction that the signup-link generation admin-panel needs to be changed to use webserver-API routes in the devops section

I do not understand this point. Could you please clarify (perhaps offline?)


> If possible: Please also add a env-var config option to the GC, by default false, to disable expired API key deletion entirely from the GC, if possible.

 why do you think this option would be useful? I am curious ... i wonder if i am missing some case here. IMO stopping this task might even lead to inconsistencies like expired tokens still active. Actually the task is intentionally designed to be resilient to failures. 


> 
> Thanks.


> Question: Will the frontend be able to display a clear "Your account is expired (please purchase more oSparc tokens)" error using this backend logic? I guess that this is at least the intuitive user-story.

Good point: short answer is no, not yet. This is a feature that is currently **only available in the API** and still not integrated in the front-end. It was intended to be used for th 5G API part of the computational services.  Nonetheles, I think is a good idea and created https://github.com/ITISFoundation/osparc-issues/issues/729

I think you got the concept wrong: there is no purchase involved here, the user creates tokens himself and gives them expiration dates.

**Review by mrnicegyu11 — APPROVED** [review]
thanks for clarifying, ok

File: services/web/server/src/simcore_service_webserver/garbage_collector_tasks_api_keys.py (hunk 1 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/garbage_collector_tasks_api_keys.py]: updated

**Review by GitHK — APPROVED** [review]
👍 Nice!

**Pull request merged**